core/src/poh_service.rs

```diff
@@ -85,27 +90,48 @@ impl PohService {
         }
     }
 
-    fn tick_producer(poh_recorder: Arc<Mutex<PohRecorder>>, poh_exit: &AtomicBool) {
+    fn tick_producer(
+        poh_recorder: Arc<Mutex<PohRecorder>>,
+        poh_exit: &AtomicBool,
+        target_tick_ns: u64,
+        ticks_per_slot: u64,
+    ) {
+        info!("starting with target ns: {}", target_tick_ns);
         let poh = poh_recorder.lock().unwrap().poh.clone();
         let mut now = Instant::now();
+        let mut last_metric = Instant::now();
         let mut num_ticks = 0;
         let mut num_hashes = 0;
+        let mut total_sleep_us = 0;
         loop {
             num_hashes += NUM_HASHES_PER_BATCH;
             if poh.lock().unwrap().hash(NUM_HASHES_PER_BATCH) {
                 // Lock PohRecorder only for the final hash...
                 poh_recorder.lock().unwrap().tick();
                 num_ticks += 1;
-                if num_ticks >= DEFAULT_TICKS_PER_SLOT * 2 {
+                let elapsed_ns = now.elapsed().as_nanos() as u64;
+                // sleep is not accurate enough to get a predictable time.
+                // Kernel can not schedule the thread for a while.
+                while (now.elapsed().as_nanos() as u64) < target_tick_ns {
+                    std::sync::atomic::spin_loop_hint();
```

Comment:
I think putting this inside the `if poh.lock().unwrap().hash(NUM_HASHES_PER_BATCH)` will hold the PoH lock for the duration of the spin, which will block others who are checking the PoH tick height

Reply:
Actually, I put a long `sleep` here and the lock is free to be taken during the sleep, so I think the lock is released after `.hash` is called.

Reply:
yup you're right!